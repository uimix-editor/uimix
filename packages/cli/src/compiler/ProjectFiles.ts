import {
  Image,
  ColorToken,
  NodeJSON,
  PageJSON,
  ProjectJSON,
  ProjectManifestJSON,
  StyleJSON,
} from "@uimix/node-data";
import { filter, omit } from "lodash-es";
import { mkdirpSync } from "mkdirp";
import { globSync } from "glob";
import path from "path";
import fs from "fs";
import { sha256 } from "js-sha256";
import { formatJSON } from "../format";
import chokidar from "chokidar";
import { encode } from "url-safe-base64";

function getURLSafeBase64Hash(data: string | Uint8Array): string {
  const hash = sha256.arrayBuffer(data);
  return encode(Buffer.from(hash).toString("base64"));
}

interface HierarchicalNodeJSON extends NodeJSON {
  id: string;
  children: HierarchicalNodeJSON[];
}

function toHierarchicalNodeJSONs(
  nodes: Record<string, NodeJSON>
): HierarchicalNodeJSON[] {
  const hierarchicalNodes: Record<string, HierarchicalNodeJSON> = {};

  for (const [id, node] of Object.entries(nodes)) {
    hierarchicalNodes[id] = {
      ...node,
      id,
      children: [],
    };
  }

  for (const [id, node] of Object.entries(nodes)) {
    if (node.parent) {
      hierarchicalNodes[node.parent].children.push(hierarchicalNodes[id]);
    }
  }

  for (const node of Object.values(hierarchicalNodes)) {
    node.children.sort((a, b) => a.index - b.index);
  }

  return Object.values(hierarchicalNodes).filter((node) => !node.parent);
}

export class ProjectFiles {
  constructor(
    rootPath: string,
    options: {
      filePattern?: string;
    } = {}
  ) {
    this.rootPath = rootPath;
    this.filePattern = options.filePattern ?? "**/*.uimix";
  }

  readonly rootPath: string;
  readonly filePattern: string;
  manifest: ProjectManifestJSON = {};
  pages = new Map<string /* file path */, PageJSON>();

  toProjectJSON(): ProjectJSON {
    // TODO: detect ID conflicts between files (and provide way to resolve them)

    const nodes: Record<string, NodeJSON> = {};
    const styles: Record<string, Partial<StyleJSON>> = {};
    const images: Record<string, Image> = {};
    const colors: Record<string, ColorToken> = {};

    const projectNode: NodeJSON = {
      type: "project",
      index: 0,
    };
    nodes["project"] = projectNode;

    let pageIndex = 0;
    for (const [pageName, pageJSON] of this.pages) {
      const pageID: string = getURLSafeBase64Hash(pageName);

      const pageNode: NodeJSON = {
        type: "page",
        index: pageIndex++,
        name: pageName,
        parent: "project",
      };
      nodes[pageID] = pageNode;

      for (const [id, node] of Object.entries(pageJSON.nodes)) {
        nodes[id] = {
          ...node,
          parent: node.parent ?? pageID,
        };
      }

      for (const [id, style] of Object.entries(pageJSON.styles)) {
        styles[id] = style;
      }
      for (const [id, image] of Object.entries(pageJSON.images ?? {})) {
        images[id] = image;
      }
      for (const [id, color] of Object.entries(pageJSON.colors ?? {})) {
        colors[id] = {
          ...color,
          page: pageID,
        };
      }
    }

    return {
      nodes,
      styles,
      componentURLs: this.manifest.componentURLs,
      images,
      colors,
    };
  }

  loadProjectJSON(projectJSON: ProjectJSON): void {
    const manifest: ProjectManifestJSON = {
      componentURLs: projectJSON.componentURLs,
    };

    const hierarchicalNodes = toHierarchicalNodeJSONs(projectJSON.nodes);
    const projectNode = hierarchicalNodes.find(
      (node) => node.type === "project"
    );
    if (!projectNode) {
      throw new Error("Project node not found");
    }

    const pages = new Map<string, PageJSON>();
    const nodeIDToPageJSON = new Map<string, PageJSON>();
    for (const page of projectNode.children) {
      if (page.type !== "page") {
        throw new Error("expected page node");
      }
      if (!page.name) {
        throw new Error("page name is empty");
      }

      const pageJSON: PageJSON = {
        nodes: {},
        styles: {},
        // TODO: selectively save images
        images: projectJSON.images,
        colors: Object.fromEntries(
          Object.entries(projectJSON.colors ?? {})
            .filter(([, color]) => color.page === page.id)
            .map(([id, color]) => [id, omit(color, ["page"])])
        ),
      };
      const addNodeRecursively = (node: HierarchicalNodeJSON) => {
        pageJSON.nodes[node.id] = omit(node, ["children", "id"]);
        nodeIDToPageJSON.set(node.id, pageJSON);
        for (const child of node.children) {
          addNodeRecursively(child);
        }
      };
      for (const child of page.children) {
        addNodeRecursively(omit(child, ["parent"]));
      }

      pages.set(page.name, pageJSON);
    }

    for (const [id, style] of Object.entries(projectJSON.styles)) {
      const idPath = id.split(":");
      const pageJSON = nodeIDToPageJSON.get(idPath[0]);
      if (pageJSON) {
        pageJSON.styles[id] = style;
      }
    }

    this.manifest = manifest;
    this.pages = pages;
  }

  load(): void {
    const rootPath = this.rootPath;
    const manifestPath = path.resolve(rootPath, "uimix/project.json");
    let manifest: ProjectManifestJSON;
    try {
      manifest = ProjectManifestJSON.parse(
        JSON.parse(fs.readFileSync(manifestPath, { encoding: "utf-8" }))
      );
    } catch {
      manifest = { componentURLs: [] };
    }

    const pages = new Map<string, PageJSON>();

    const pagePaths = globSync(this.filePattern, {
      cwd: rootPath,
    });
    pagePaths.sort();

    for (const pagePath of pagePaths) {
      const pageJSON = PageJSON.parse(
        JSON.parse(
          fs.readFileSync(path.resolve(rootPath, pagePath), {
            encoding: "utf-8",
          })
        )
      );
      pages.set(pagePath.replace(/\.uimix$/, ""), pageJSON);
    }

    this.manifest = manifest;
    this.pages = pages;
  }

  save(): void {
    const rootPath = this.rootPath;

    mkdirpSync(path.resolve(rootPath, "uimix"));
    mkdirpSync(path.resolve(rootPath, "uimix/images"));

    fs.writeFileSync(
      path.resolve(rootPath, "uimix/project.json"),
      formatJSON(JSON.stringify(this.manifest))
    );

    // for (const [hash, image] of Object.entries(this.manifest.images ?? {})) {
    //   const dataURL = image.url;
    //   // data url to buffer
    //   const base64 = dataURL.split(",")[1];
    //   const buffer = Buffer.from(base64, "base64");
    //   const ext = mime.extension(image.type);
    //   if (!ext) {
    //     console.error(`Unknown MIME type for ${image.type}`);
    //     continue;
    //   }

    //   fs.writeFileSync(
    //     path.resolve(rootPath, `uimix/images/${hash}.${ext}`),
    //     buffer
    //   );
    // }

    const pagePathsToDelete = new Set(
      globSync(this.filePattern, {
        cwd: rootPath,
      })
    );

    for (const [pageName, pageJSON] of this.pages) {
      const pagePath = pageName + ".uimix";
      const pageDirPath = path.dirname(pagePath);
      mkdirpSync(path.resolve(rootPath, pageDirPath));
      fs.writeFileSync(
        path.resolve(rootPath, pagePath),
        formatJSON(JSON.stringify(pageJSON))
      );
      pagePathsToDelete.delete(pagePath);
    }

    for (const pagePath of pagePathsToDelete) {
      fs.rmSync(path.resolve(rootPath, pagePath));
    }
  }

  watch(onChange: () => void): () => void {
    const watchPath = path.resolve(this.rootPath, this.filePattern);

    // FIXME: chokidar looks like making UI slow
    const watcher = chokidar.watch(watchPath, {
      ignored: ["**/node_modules/**", "**/dist/**"],
    });
    console.log("start watching...");

    const _onChange = () => {
      try {
        this.load();
        onChange();
      } catch (e) {
        console.error(e);
      }
    };

    watcher.on("change", _onChange);
    watcher.on("add", _onChange);
    watcher.on("unlink", _onChange);

    return () => watcher.close();
  }
}
