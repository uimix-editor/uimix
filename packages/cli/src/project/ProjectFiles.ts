import {
  Image,
  ColorToken,
  NodeJSON,
  PageJSON,
  ProjectJSON,
  ProjectManifestJSON,
  StyleJSON,
} from "@uimix/model/src/data/v1";
import { getPageID, compareProjectJSONs } from "@uimix/model/src/data/util";
import { omit } from "lodash-es";
import { formatJSON } from "../format";
import { FileAccess } from "./FileAccess";

interface HierarchicalNodeJSON extends NodeJSON {
  id: string;
  children: HierarchicalNodeJSON[];
}

function toHierarchicalNodeJSONs(
  nodes: Record<string, NodeJSON>
): HierarchicalNodeJSON[] {
  const hierarchicalNodes: Record<string, HierarchicalNodeJSON> = {};

  for (const [id, node] of Object.entries(nodes)) {
    hierarchicalNodes[id] = {
      ...node,
      id,
      children: [],
    };
  }

  for (const [id, node] of Object.entries(nodes)) {
    if (node.parent) {
      hierarchicalNodes[node.parent].children.push(hierarchicalNodes[id]);
    }
  }

  for (const node of Object.values(hierarchicalNodes)) {
    node.children.sort((a, b) => a.index - b.index);
  }

  return Object.values(hierarchicalNodes).filter((node) => !node.parent);
}

interface ProjectFilesOptions {
  filePattern?: string;
}

// Important TODO: fix paths in Windows!!
export class ProjectFiles {
  static async load(fileAccess: FileAccess, options: ProjectFilesOptions = {}) {
    const projectFiles = new ProjectFiles(fileAccess, options);
    await projectFiles.load();
    return projectFiles;
  }

  constructor(fileAccess: FileAccess, options: ProjectFilesOptions = {}) {
    this.fileAccess = fileAccess;
    this.filePattern = options.filePattern ?? "**/*.uimix";
  }

  readonly fileAccess: FileAccess;

  get rootPath(): string {
    return this.fileAccess.rootPath;
  }

  readonly filePattern: string;
  readonly manifestName = "uimix.json";
  json: ProjectJSON = {
    nodes: {},
    styles: {},
  };

  async load(): Promise<boolean> {
    let manifest: ProjectManifestJSON;
    try {
      manifest = ProjectManifestJSON.parse(
        JSON.parse(await this.fileAccess.readText(this.manifestName))
      );
    } catch {
      manifest = { componentURLs: [] };
    }

    const pages = new Map<string, PageJSON>();

    const pagePaths = await this.fileAccess.glob(this.filePattern);
    pagePaths.sort();

    for (const pagePath of pagePaths) {
      const pageJSON = PageJSON.parse(
        JSON.parse(await this.fileAccess.readText(pagePath))
      );
      pages.set(pagePath.replace(/\.uimix$/, ""), pageJSON);
    }
    const newProjectJSON = filesToProjectJSON(manifest, pages);

    if (compareProjectJSONs(this.json, newProjectJSON)) {
      return false;
    }

    this.json = newProjectJSON;
    return true;
  }

  async save(): Promise<void> {
    const { manifest, pages } = projectJSONToFiles(this.json);

    await this.fileAccess.writeText(
      this.manifestName,
      formatJSON(JSON.stringify(manifest))
    );

    const pagePathsToDelete = new Set(
      await this.fileAccess.glob(this.filePattern)
    );

    for (const [pageName, pageJSON] of pages) {
      const pagePath = pageName + ".uimix";
      await this.fileAccess.writeText(
        pagePath,
        formatJSON(JSON.stringify(pageJSON))
      );
      pagePathsToDelete.delete(pagePath);
    }

    for (const pagePath of pagePathsToDelete) {
      await this.fileAccess.remove(pagePath);
    }
  }

  watch(onChange: (projectJSON: ProjectJSON) => void): () => void {
    console.log("start watching...");

    return this.fileAccess.watch(this.filePattern, async () => {
      try {
        if (await this.load()) {
          onChange(this.json);
        }
      } catch (e) {
        console.error(e);
      }
    });
  }
}

export function projectJSONToFiles(projectJSON: ProjectJSON): {
  manifest: ProjectManifestJSON;
  pages: Map<string, PageJSON>;
} {
  const manifest: ProjectManifestJSON = {
    componentURLs: projectJSON.componentURLs,
  };

  const hierarchicalNodes = toHierarchicalNodeJSONs(projectJSON.nodes);
  const projectNode = hierarchicalNodes.find((node) => node.type === "project");
  if (!projectNode) {
    throw new Error("Project node not found");
  }

  const pages = new Map<string, PageJSON>();
  const nodeIDToPageJSON = new Map<string, PageJSON>();
  for (const page of projectNode.children) {
    if (page.type !== "page") {
      throw new Error("expected page node");
    }
    if (!page.name) {
      throw new Error("page name is empty");
    }

    const pageJSON: PageJSON = {
      nodes: {},
      styles: {},
      // TODO: selectively save images
      images: projectJSON.images,
      colors: Object.fromEntries(
        Object.entries(projectJSON.colors ?? {})
          .filter(([, color]) => color.page === page.id)
          .map(([id, color]) => [id, omit(color, ["page"])])
      ),
    };
    const addNodeRecursively = (node: HierarchicalNodeJSON) => {
      pageJSON.nodes[node.id] = omit(node, ["children", "id"]);
      nodeIDToPageJSON.set(node.id, pageJSON);
      for (const child of node.children) {
        addNodeRecursively(child);
      }
    };
    for (const child of page.children) {
      addNodeRecursively(omit(child, ["parent"]));
    }

    pages.set(page.name, pageJSON);
  }

  for (const [id, style] of Object.entries(projectJSON.styles)) {
    const idPath = id.split(":");
    const pageJSON = nodeIDToPageJSON.get(idPath[0]);
    if (pageJSON) {
      pageJSON.styles[id] = style;
    }
  }

  return {
    manifest,
    pages,
  };
}

export function filesToProjectJSON(
  manifest: ProjectManifestJSON,
  pages: Map<string, PageJSON>
): ProjectJSON {
  // TODO: detect ID conflicts between files (and provide way to resolve them)

  const nodes: Record<string, NodeJSON> = {};
  const styles: Record<string, Partial<StyleJSON>> = {};
  const images: Record<string, Image> = {};
  const colors: Record<string, ColorToken> = {};

  const projectNode: NodeJSON = {
    type: "project",
    index: 0,
  };
  nodes["project"] = projectNode;

  let pageIndex = 0;
  for (const [pageName, pageJSON] of pages) {
    const pageID: string = getPageID(pageName);

    const pageNode: NodeJSON = {
      type: "page",
      index: pageIndex++,
      name: pageName,
      parent: "project",
    };
    nodes[pageID] = pageNode;

    for (const [id, node] of Object.entries(pageJSON.nodes)) {
      nodes[id] = {
        ...node,
        parent: node.parent ?? pageID,
      };
    }

    for (const [id, style] of Object.entries(pageJSON.styles)) {
      styles[id] = style;
    }
    for (const [id, image] of Object.entries(pageJSON.images ?? {})) {
      images[id] = image;
    }
    for (const [id, color] of Object.entries(pageJSON.colors ?? {})) {
      colors[id] = {
        ...color,
        page: pageID,
      };
    }
  }

  return {
    nodes,
    styles,
    componentURLs: manifest.componentURLs,
    images,
    colors,
  };
}
