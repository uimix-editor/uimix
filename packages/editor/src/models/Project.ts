/* eslint-disable @typescript-eslint/no-explicit-any */

import * as Y from "yjs";
import { posix as path } from "path-browserify";
import { getOrCreate } from "../state/Collection";
import { SelectableMap } from "./Selectable";
import { Node, NodeMap } from "./Node";
import { computed, makeObservable, ObservableMap } from "mobx";
import { ProjectJSON } from "@uimix/node-data";
import { toProjectJSON } from "./toProjectJSON";
import { ImageManager } from "./ImageManager";
import { Component } from "./Component";
import { ObservableYMap } from "../utils/ObservableYMap";

export interface PageHierarchyFolderEntry {
  type: "directory";
  path: string;
  name: string;
  children: PageHierarchyEntry[];
}

export interface PageHierarchyPageEntry {
  type: "file";
  path: string;
  name: string;
  page: Node;
}

export type PageHierarchyEntry =
  | PageHierarchyFolderEntry
  | PageHierarchyPageEntry;

class Pages {
  constructor(project: Project) {
    this.project = project;
    this.node = project.node;
    makeObservable(this);
  }

  node: Node;
  readonly project: Project;

  @computed get count(): number {
    return this.node.childCount;
  }

  get all(): Node[] {
    return this.node.children;
  }

  create(filePath: string): Node {
    const node = this.project.nodes.create("page");
    node.name = filePath;
    return node;
  }

  toHierarchy(): PageHierarchyFolderEntry {
    const root: PageHierarchyFolderEntry = {
      type: "directory",
      name: "",
      path: "",
      children: [],
    };
    const parents = new Map<string, PageHierarchyFolderEntry>();
    parents.set("", root);

    const mkdirp = (segments: string[]): PageHierarchyFolderEntry => {
      if (segments.length === 0) {
        return root;
      }

      const existing = parents.get(segments.join("/"));
      if (existing) {
        return existing;
      }

      const parent = mkdirp(segments.slice(0, -1));
      const dir: PageHierarchyFolderEntry = {
        type: "directory",
        name: segments[segments.length - 1],
        path: segments.join("/"),
        children: [],
      };
      parent.children.push(dir);
      parents.set(segments.join("/"), dir);
      return dir;
    };

    const pages = Array.from(this.all);
    pages.sort((a, b) => a.name.localeCompare(b.name));

    for (const page of pages) {
      const segments = page.name.split(path.sep);
      const parent = mkdirp(segments.slice(0, -1));

      const item: PageHierarchyPageEntry = {
        type: "file",
        name: segments[segments.length - 1],
        path: page.name,
        page,
      };
      parent.children.push(item);
    }

    return root;
  }

  affectedPagesForPath(path: string): Node[] {
    return this.all.filter(
      (page) => page.name === path || page.name.startsWith(path + "/")
    );
  }

  deletePageOrPageFolder(path: string) {
    const pagesToDelete = this.affectedPagesForPath(path);

    for (const page of pagesToDelete) {
      page.remove();
      // TODO: delete dangling nodes?
      //this.project.nodes.remove(doc.root);
    }
  }

  renamePageOrPageFolder(path: string, newPath: string) {
    if (path === newPath) {
      return;
    }

    const pagesToDelete = this.affectedPagesForPath(path);

    for (const page of pagesToDelete) {
      page.name = newPath + page.name.slice(path.length);
    }
  }
}

export class Project {
  constructor(data: Y.Map<any>) {
    this.data = ObservableYMap.get(data);
    this.nodes = new NodeMap(this);
    this.selectables = new SelectableMap(this);
    this.node = this.nodes.create("project", "project");
    this.pages = new Pages(this);
  }

  readonly data: ObservableYMap<any>;
  readonly nodes: NodeMap;
  readonly selectables: SelectableMap;
  readonly node: Node;
  readonly pages: Pages;
  readonly imageManager = new ImageManager();

  toJSON(): ProjectJSON {
    return toProjectJSON(this.node.children.map((c) => c.selectable));
  }

  loadJSON(json: ProjectJSON) {
    this.node.clear();

    for (const [id, nodeJSON] of Object.entries(json.nodes)) {
      const node = this.nodes.getOrCreate(nodeJSON.type, id);
      node.loadJSON(nodeJSON);
      if (node.type === "page") {
        this.node.append([node]);
      }
    }
    for (const [id, style] of Object.entries(json.styles)) {
      const selectable = this.selectables.get(id.split(":"));
      selectable.selfStyle.loadJSON(style);
    }
  }

  get components(): Component[] {
    const components: Component[] = [];

    for (const page of this.pages.all) {
      for (const node of page.children) {
        const component = Component.from(node);
        if (component) {
          components.push(component);
        }
      }
    }

    return components;
  }
}
