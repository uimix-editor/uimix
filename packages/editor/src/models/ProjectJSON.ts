import { ProjectJSON } from "@uimix/node-data";
import * as Y from "yjs";

export function loadProjectJSON(ydoc: Y.Doc, projectJSON: ProjectJSON): void {
  // TODO: support overwrite

  const nodes = ydoc.getMap("nodes");
  for (const [id, json] of Object.entries(projectJSON.nodes)) {
    const data = new Y.Map();
    for (const [key, value] of Object.entries(json)) {
      data.set(key, value);
    }
    nodes.set(id, data);
  }

  const styles = ydoc.getMap("styles");
  for (const [id, json] of Object.entries(projectJSON.styles)) {
    const data = new Y.Map();
    for (const [key, value] of Object.entries(json)) {
      data.set(key, value);
    }
    styles.set(id, data);
  }

  const componentURLs = ydoc.getArray("componentURLs");
  componentURLs.delete(0, componentURLs.length);
  for (const url of projectJSON.componentURLs ?? []) {
    componentURLs.push([url]);
  }

  const images = ydoc.getMap("images");
  images.clear();
  for (const [hash, url] of Object.entries(projectJSON.images ?? {})) {
    images.set(hash, url);
  }
}

export function toProjectJSON(ydoc: Y.Doc): ProjectJSON {
  const json: ProjectJSON = {
    nodes: ydoc.getMap("nodes").toJSON(),
    styles: ydoc.getMap("styles").toJSON(),
    componentURLs: ydoc.getArray("componentURLs").toJSON(),
    images: ydoc.getMap("images").toJSON(),
  };
  // delete dangling nodes
  for (const id of Object.keys(json.nodes)) {
    if (json.nodes[id].type !== "project" && !json.nodes[id].parent) {
      delete json.nodes[id];
    }
  }
  // delete styles for deleted nodes
  for (const styleID of Object.keys(json.styles)) {
    const ids = styleID.split(":");
    for (const id of ids) {
      if (!json.nodes[id]) {
        delete json.styles[styleID];
      }
    }
  }
  return json;
}
