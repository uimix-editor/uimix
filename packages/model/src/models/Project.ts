import * as Y from "yjs";
import { Selectable, SelectableMap } from "./Selectable";
import { Node, NodeMap } from "./Node";
import * as Data from "../data/v1";
import { ImageManager } from "./ImageManager";
import { Component } from "./Component";
import { ObservableYArray } from "@uimix/foundation/src/utils/ObservableYArray";
import { ColorTokenMap } from "./ColorToken";
import { PageList } from "./PageList";
import { ProjectData } from "../collaborative/ProjectData";
import { CodeAssets } from "./CodeAssets";
import { makeObservable, observable } from "mobx";
import { Page } from "./Page";

export class Project {
  constructor() {
    this.nodes = new NodeMap(this);
    this.selectables = new SelectableMap(this);
    this.colorTokens = new ColorTokenMap(this);
    this.node = this.nodes.create("project", "project");
    this.pages = new PageList(this);
    makeObservable(this);
  }

  readonly data = new ProjectData();
  readonly nodes: NodeMap;
  readonly selectables: SelectableMap;
  readonly colorTokens: ColorTokenMap;
  readonly node: Node;
  readonly pages: PageList;
  readonly imageManager = new ImageManager(this);

  get doc(): Y.Doc {
    return this.data.doc;
  }

  createUndoManager(): Y.UndoManager {
    return new Y.UndoManager([
      this.data.nodes,
      this.data.styles,
      this.data.selection,
      this.data.componentURLs,
      this.data.colors,
    ]);
  }

  clear(): void {
    this.data.clear();
    this.nodes.create("project", "project");
  }

  toJSON(): Data.Project {
    return this.data.toJSON();
  }

  loadJSON(json: Data.Project) {
    this.data.loadJSON(json);
  }

  get components(): Component[] {
    return this.pages.all.flatMap((page) => page.components);
  }

  get componentURLs(): ObservableYArray<string> {
    return ObservableYArray.get(this.data.componentURLs);
  }

  @observable localCodeAssets: CodeAssets | undefined = undefined;

  // Selection

  clearSelection() {
    this.selectables.selectionData.clear();
  }

  replaceSelection(selectables: Iterable<Selectable>) {
    this.clearSelection();
    for (const selectable of selectables) {
      selectable.select();
    }
  }

  // Get by ID

  nodeForID(id: string): Node | undefined {
    return this.nodes.get(id);
  }

  pageForID(id: string): Page | undefined {
    const node = this.nodeForID(id);
    if (node) {
      return Page.from(node);
    }
  }

  componentForID(id: string): Component | undefined {
    const node = this.nodeForID(id);
    if (node) {
      return Component.from(node);
    }
  }
}
