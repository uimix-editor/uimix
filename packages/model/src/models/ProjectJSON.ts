import { ProjectJSON } from "../data/v1";
import { isEqual, omit } from "lodash-es";
import sha256 from "crypto-js/sha256";
import encBase64url from "crypto-js/enc-base64url";

export function getPageID(pageName: string): string {
  return sha256(pageName).toString(encBase64url);
}

export function compareProjectJSONs(a: ProjectJSON, b: ProjectJSON): boolean {
  // skipping images as they may be very large
  return isEqual(omit(a, "images"), omit(b, "images"));
}
