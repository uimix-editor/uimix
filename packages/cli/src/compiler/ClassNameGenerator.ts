import { Project } from "@uimix/model/src/models";
import {
  IncrementalUniqueNameGenerator,
  generateUpperJSIdentifier,
} from "@uimix/foundation/src/utils/Name";

export class ClassNameGenerator {
  constructor(project: Project) {
    this.project = project;

    for (const page of project.pages.all) {
      const namegen = new IncrementalUniqueNameGenerator();

      const pageID = page.id.slice(0, 6);

      for (const component of page.components) {
        const componentName = namegen.generate(
          generateUpperJSIdentifier(component.name)
        );

        const refIDs = component.refIDs;
        for (const [nodeID, refID] of refIDs) {
          const className = `${componentName}${pageID}__${refID}`;
          this.uniqueNames.set(nodeID, className);
        }

        this.uniqueNames.set(
          component.rootNode.id,
          `${componentName}${pageID}`
        );
      }
    }
  }

  getUniqueName(nodeID: string): string {
    const className = this.uniqueNames.get(nodeID);
    if (!className) {
      throw new Error(`No class name for node ${nodeID}`);
    }
    return className;
  }

  get(nodePath: string[]) {
    const readableIDPath = nodePath.map((id) => this.getUniqueName(id));
    return `uimix-${readableIDPath.join("-")}`;
  }

  project: Project;
  uniqueNames = new Map<string /* node id */, string>();
}
