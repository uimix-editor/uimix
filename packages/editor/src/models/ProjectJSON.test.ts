import { describe, expect, it } from "vitest";
import * as Y from "yjs";
import { loadProjectJSON, toProjectJSON } from "./ProjectJSON";
import { Project } from "./Project";
import * as path from "path";
import { ProjectFiles } from "../../../cli/src/compiler/ProjectFiles";

describe(loadProjectJSON.name, () => {
  it("works", () => {
    const rootDir = path.resolve("../sandbox");
    const projectFiles = new ProjectFiles(rootDir, {
      filePattern: "src/components.uimix",
    });
    projectFiles.load();
    const projectJSON = projectFiles.projectJSON;

    const ydoc = new Y.Doc();
    const project = new Project(ydoc);
    loadProjectJSON(ydoc, projectJSON);

    expect([...project.nodes.data.keys()]).toMatchSnapshot();
    expect([...project.selectables.stylesData.keys()]).toMatchSnapshot();
    expect(project.imageManager.images.toJSON()).toMatchSnapshot();
    expect(project.componentURLs).toMatchSnapshot();
    expect(toProjectJSON(ydoc)).toEqual(projectJSON);
  });
});
