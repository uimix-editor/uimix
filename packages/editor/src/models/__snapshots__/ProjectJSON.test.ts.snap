// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadProjectJSON > works 1`] = `
[
  "project",
  "1Bvi2ymHPmVHms6RDurmNN",
  "1Bvi2z3USBZi1pWrvPd6Ut",
  "1Bvi2z3UPfi6W3L8nDt5VA",
  "1Bvi2z3UGTM64JtRT7Qc1S",
  "1Bvi2z3UVsEskiSt2TcdL7",
  "1Bvi2z3UaHFmaikKtfcmGX",
  "1Bvi2zJALza4DrqmNaGYAk",
  "1Bvi2z3TfkSpYXW29daTN8",
  "1Bvi2z3TpKXtTJLTRPTEd8",
  "1Bvi2z3To8ddXo6R6ZftVD",
  "1Bvi2z3UB1xUVg4Xz9WTpJ",
  "1Bvi2z3U3SiK6r7gf9vh7N",
  "1Bvi3BSSb2oaHb1ZqUEk6H",
  "1Bvi319YVxZELz4tEskN1S",
  "1Bvi31VKu8gfy6XzJ8xrHA",
  "1Bvi365mtWZmXKDPeJTB4s",
  "1Bvi32d9RMi3ES5kk2UvuM",
  "1Bvi2z3UrU9VkawwAteXs8",
  "1Bvi392qBdMy93LGLcVymb",
  "1Bvi39mct9UNYZ9eTCCRdV",
]
`;

exports[`loadProjectJSON > works 2`] = `
[
  "1Bvi2ymHPmVHms6RDurmNN",
  "1Bvi2z3USBZi1pWrvPd6Ut",
  "1Bvi2z3UPfi6W3L8nDt5VA",
  "1Bvi2z3UGTM64JtRT7Qc1S",
  "1Bvi2z3UVsEskiSt2TcdL7",
  "1Bvi2z3UVsEskiSt2TcdL7:1Bvi2z3UGTM64JtRT7Qc1S",
  "1Bvi2z3UaHFmaikKtfcmGX",
  "1Bvi2z3UaHFmaikKtfcmGX:1Bvi2z3UGTM64JtRT7Qc1S",
  "1Bvi2zJALza4DrqmNaGYAk",
  "1Bvi2z3TfkSpYXW29daTN8",
  "1Bvi2z3TpKXtTJLTRPTEd8",
  "1Bvi2z3To8ddXo6R6ZftVD",
  "1Bvi2z3UB1xUVg4Xz9WTpJ",
  "1Bvi2z3U3SiK6r7gf9vh7N",
  "1Bvi3BSSb2oaHb1ZqUEk6H",
  "1Bvi319YVxZELz4tEskN1S",
  "1Bvi31VKu8gfy6XzJ8xrHA",
  "1Bvi365mtWZmXKDPeJTB4s",
  "1Bvi32d9RMi3ES5kk2UvuM",
  "1Bvi2z3UrU9VkawwAteXs8",
  "1Bvi2z3UrU9VkawwAteXs8:1Bvi2z3UGTM64JtRT7Qc1S",
  "1Bvi392qBdMy93LGLcVymb",
  "1Bvi39mct9UNYZ9eTCCRdV",
  "1Bvi39mct9UNYZ9eTCCRdV:1Bvi2z3TpKXtTJLTRPTEd8",
  "1Bvi39mct9UNYZ9eTCCRdV:1Bvi2z3To8ddXo6R6ZftVD",
  "1Bvi39mct9UNYZ9eTCCRdV:1Bvi2z3UB1xUVg4Xz9WTpJ",
  "1Bvi39mct9UNYZ9eTCCRdV:1Bvi2z3U3SiK6r7gf9vh7N",
]
`;
